src/main/java/com/amadeus/jenkins/plugins/workflow/libs/HttpRetriever.java

```diff
@@ -241,6 +244,12 @@ private FilePath download(String sourceURL, UsernamePasswordCredentials password
         try (CloseableHttpClient client = HttpClients.createDefault()) {
             HttpClientContext context = getHttpClientContext(passwordCredentials);
             HttpGet get = new HttpGet(new URL(sourceURL).toURI());
+            if (passwordCredentials != null) {
```

Comment:
The proposed change sounds good. I just need to double check two things now:
* if it still works in fine in my case (an additional non regression test)
* why the call to setCredentialsProvider(CredentialsProvider credentialsProvider) is not enough

Reply:
I think Artifactory behaves a bit special here.
If anonymous users are enabled then Artifactory interprets a request without authentication as an anonymous user. Which may have not enough permissions.

IMO a better way than enabling this unconditionally would be with an explicit setting `preemptiveAuthentication`.

A more elegant implementation would maybe be
```
context.getTargetAuthState().setState(AuthProtocolState.CHALLENGED);
```
(Untested, there are also other ways of making the http library unconditionally send the authentication)

Reply:
In my testing, the existing code works fine with a `maven` repository with *force authentication* `enabled`. In my case the jenkins shared library was in a `generic` repo. I have anonymous auth enabled for some repositories but not for the one that had the shared jenkins library. Artifactory returns 400 error in such cases, hence my change. Hope this clarifies.

Reply:
Let me know what I need to get this going forward. The jenkins version change was so that `mvn hpi:run` actually worked. It didn't work for me with the old version. I can revert that back happily 👍

Reply:
@farshidce @t-8ch any suggestions?

I tried the code suggested by @t-8ch and it seems not to be working.

Reply:
I did not have the time to dive in your issue yet. But you're clearly on my todo list. I come back to you as soon as I can. Probably during next week.

Reply:
This should work: https://hc.apache.org/httpcomponents-client-ga/httpclient/examples/org/apache/http/examples/client/ClientPreemptiveBasicAuthentication.java
It should really not be the default though, and be tested.